array_api_strict/_data_type_functions.py

```diff
@@ -111,22 +111,32 @@ class finfo_object:
     smallest_normal: float
     dtype: DType
 
+    __hash__ = NotImplemented
 
-@dataclass
+
+@dataclass(frozen=True, eq=False)
 class iinfo_object:
     bits: int
     max: int
     min: int
     dtype: DType
 
+    __hash__ = NotImplemented
+
 
 def finfo(type: DType | Array, /) -> finfo_object:
     """
     Array API compatible wrapper for :py:func:`np.finfo <numpy.finfo>`.
 
     See its docstring for more information.
     """
-    np_type = type._array if isinstance(type, Array) else type._np_dtype
+    if isinstance(type, Array):
+        np_type = type._dtype._np_dtype
```

Comment:
Note that this should be tested in the Array API tests, not here

Reply:
Do I take it you're up to adding that test https://github.com/data-apis/array-api-tests/blob/c48410f96fc58e02eea844e6b7f6cc01680f77ce/array_api_tests/test_data_type_functions.py#L151 :-).